solidity/contracts/bridge/library/FraudVerification.sol

```diff
@@ -0,0 +1,285 @@
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+
+pragma solidity 0.8.4;
+
+import {BytesLib} from "@keep-network/bitcoin-spv-sol/contracts/BytesLib.sol";
+import {BTCUtils} from "@keep-network/bitcoin-spv-sol/contracts/BTCUtils.sol";
+import {
+    CheckBitcoinSigs
+} from "@keep-network/bitcoin-spv-sol/contracts/CheckBitcoinSigs.sol";
+import "../Bridge.sol";
+
+library FraudVerification {
+    using BytesLib for bytes;
+    using BTCUtils for bytes;
+    using BTCUtils for uint32;
+
+    struct Parameters {
+        /// The amount of ETH the party challenging the wallet for fraud needs
+        /// to deposit.
+        uint256 fraudChallengeDepositAmount; //TODO: Initialize
+    }
+
+    struct FraudChallenge {
+        address challenger;
+        uint256 ethDepositAmount;
+        uint32 reportedAt;
+        bool defended;
+    }
+
+    struct Data {
+        Parameters parameters;
+        /// Collection of all submitted fraud challenges indexed by challenge
+        /// key built as keccak256(walletPublicKey|sighash|v|r|s).
+        mapping(uint256 => FraudChallenge) fraudChallenges;
+    }
+
+    event FraudChallengeDepositAmountUpdated(
+        uint256 newFraudChallengeDepositAmount
+    );
+
+    event FraudChallengeSubmitted(
+        bytes walletPublicKey,
+        bytes32 sighash,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    );
+
+    event FraudChallengeDefended(
+        bytes walletPublicKey,
+        bytes32 sighash,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    );
+
+    /// @notice Submits a wallet fraud challenge.
+    /// @param walletPublicKey The public key of the wallet in the uncompressed
+    ///                        and unprefixed format (64 bytes).
+    /// @param sighash Hash of the data (preimage) used for generating signature
+    ///                for the given input.
+    /// @param v Signature recovery value.
+    /// @param r Signature r value.
+    /// @param s Signature s value.
+    function submitFraudChallenge(
+        Data storage data,
+        bytes memory walletPublicKey,
+        bytes32 sighash,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external {
+        // TODO: Bitcoin uses signatures in the DER format. From a signature in
+        // DER form values r and s can be easily extracted. Check how to obtain
+        // the v value. It can take values: 27, 28, 29, 30, 31, 32, 33, 34.
+        // https://bitcoin.stackexchange.com/questions/38351/ecdsa-v-r-s-what-is-v
+        require(
+            msg.value >= data.parameters.fraudChallengeDepositAmount,
+            "The amount of ETH deposited is too low"
+        );
+
+        bool verificationResult =
+            CheckBitcoinSigs.checkSig(walletPublicKey, sighash, v, r, s);
+        require(verificationResult, "Signature verification failure");
+
```

Comment:
We can inline the `checkSig` call within the `require`. I think `verificationResult` is redundant here.

Reply:
Done.